src/System.Linq.Expressions/tests/ILReader/LocalsSignatureParser.cs

```diff
@@ -0,0 +1,335 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace System.Linq.Expressions.Tests
+{
+    internal class LocalsSignatureParser : SigParser
+    {
+        private readonly Stack<Type> _types = new Stack<Type>();
+        private readonly Stack<TypeCreator> _creators = new Stack<TypeCreator>();
+        private readonly ITokenResolver _resolver;
+
+        public LocalsSignatureParser(ITokenResolver resolver)
+        {
+            _resolver = resolver;
+        }
+
+        public bool Parse(byte[] sig, out Type[] types)
+        {
+            if (Parse(sig))
+            {
+                types = _types.Reverse().ToArray();
+                return true;
+            }
+
+            types = null;
+            return false;
+        }
+
+        private static Type ToType(byte corType)
+        {
+            switch (corType)
+            {
+                case ELEMENT_TYPE_VOID: return typeof(void);
+                case ELEMENT_TYPE_BOOLEAN: return typeof(bool);
+                case ELEMENT_TYPE_CHAR: return typeof(char);
+                case ELEMENT_TYPE_I1: return typeof(sbyte);
+                case ELEMENT_TYPE_U1: return typeof(byte);
+                case ELEMENT_TYPE_I2: return typeof(short);
+                case ELEMENT_TYPE_U2: return typeof(ushort);
+                case ELEMENT_TYPE_I4: return typeof(int);
+                case ELEMENT_TYPE_U4: return typeof(uint);
+                case ELEMENT_TYPE_I8: return typeof(long);
+                case ELEMENT_TYPE_U8: return typeof(ulong);
+                case ELEMENT_TYPE_R4: return typeof(float);
+                case ELEMENT_TYPE_R8: return typeof(double);
+                case ELEMENT_TYPE_STRING: return typeof(string);
+                case ELEMENT_TYPE_I: return typeof(IntPtr);
+                case ELEMENT_TYPE_U: return typeof(UIntPtr);
+                case ELEMENT_TYPE_OBJECT: return typeof(object);
+            }
+
+            throw new NotSupportedException();
+        }
+
+        class TypeCreator
+        {
+            private static readonly MethodInfo s_GetTypeFromHandleUnsafe = typeof(Type).GetMethodAssert("GetTypeFromHandleUnsafe");
```

Comment:
I see... I'm willing to pursue this route; do you expect pushback to adding this to the model (it'd remain `internal` either way)?

In the short term, I can work around it to unblock this type of testing here. Most types that end up in locals are either on a short list of helper types used by the compiler (e.g. `StrongBox<>`) or get mentioned within the tree, so I can collect a list during a pre-analysis visitor pass over the tree, and build a map from `IntPtr` to `Type` because we have `Value` exposed on `RuntimeTypeHandle`.